### ISSUE TITLE
Coverage not generating when tests run with TypeScript + ES Modules + ts-node/esm Loader

### ISSUE THREAD

**Issue by chill-cod3r** [tapjs/tapjs#813]
Hi, first off, I want to say this testing library is amazing - by far my favorite testing library for node.js 🎉 .  

With that, I feel like I may have found an obscure scenario when trying to refactor an existing TypeScript codebase to use ES Modules.  I'm finding that coverage no longer generates in the console.  I'm guessing it could be a bug in the dependency (nyc I believe?) but I'm not sure how to tell.  I whipped up a quick repo with the `fastify-cli` to demonstrate:

Running just with tap:
`npx tap --node-arg=--loader=ts-node/esm test/**/*.test.ts`
<img width="1133" alt="image" src="https://user-images.githubusercontent.com/22133869/157788858-71e4f4ca-e77c-4b69-89a9-6e0e4fc8b6b4.png">

However (and this is what leads me to believe it's a dependency issue) - if I pipe the coverage into something like [c8](https://www.npmjs.com/package/c8) - c8 is able to seemingly generate coverage with no issue (which makes me think that tap isn't doing anything wrong here).  See:
`npx c8 tap --reporter=tap --node-arg=--loader=ts-node/esm test/**/*.test.ts`
<img width="608" alt="image" src="https://user-images.githubusercontent.com/22133869/157789019-14655941-abc4-480c-bca5-f6438bfd2c57.png">

Here is the repo (hopefully it helps):
https://github.com/wolfejw86/node-tap-coverage-possible-bug 
Any advice on how to proceed, or maybe even if you have a suggestion of another "thread to pull" - I'm happy to help.

Thanks!

EDIT: Possibly related to this: https://github.com/istanbuljs/nyc/issues/1343 - similar issue to trying to use the `mock-import` package - ie "just use this loader and it will all work" - however afaik with node.js you can only use a single loader at a time - so perhaps a fix could be to write a custom loader that handles ts-node/esm, mock-import, and @istanbul/loader - seems like a lot and I am getting near the edge of my understanding of ESM issues at this point for sure 😅

**Comment by isaacs** [comment]
Not exactly a duplicate of #807, but related, and will be fixed by the same thing that fixes it.  Work is underway at https://github.com/tapjs/processinfo

**Comment by isaacs** [comment]
tl;dr - the module loading hooks that istanbul uses don't work with the new stuff in ESM.  There's a loader API, and there _used to_ be a `--loader` you could provide that would dtrt ([@istanbuljs/esm-loader-hook](https://www.npmjs.com/package/@istanbuljs/esm-loader-hook)), but then the loader API changed, and there's no one to keep maintaining @istanbuljs/esm-loader-hook.

So my plan, now that ESM loader API is settling somewhat (though still "experimental") is to just rip out NYC entirely, use ESM loaders all the way down.  I'll hang coverage and processinfo on that, and be able to load ts modules using the `ts-node/esm` loader.

In the meantime, if your coverage needs are simple and you don't mind a lot of excessive data going into your temp dir, you can run your scripts under c8 like `c8 tap --no-cov && c8 report` and that'll _mostly_ work, or maybe get you by until tap can do it all for you again.

**Comment by chill-cod3r** [comment]
Okay cool, thank you for that explanation- I appreciate it!

**Comment by aakoch** [comment]
I'm still using `c8 tap --no-coverage` to produce coverage. This makes me think this was fixed..? I don't have an TS requirement, but found I couldn't use `--watch` because I get `Error: --watch requires coverage to be enabled`. Just a comment. I don't need a response. And thanks for making this!

**Comment by isaacs** [comment]
Yeah, c8 will provide coverage for esm (including esm+typescript).  But tap relies on nyc for processinfo tracking, so that's also gotta be rebuilt.  Work is happening over at https://github.com/tapjs/processinfo
